chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,409 @@
+;
+;   By induction, we see that S_{m+1} = sum(i, 1, m) [O_i - A_i] = sum(i, 1, m) O_i - sum(i, 1, m) A_i
+;   So S_{n+1} = sum(i, 1, n) O_i - sum(i, 1, n) A_i. But S_{n+1} is actually S_1 = 0,
+;   so thus sum(i, 1, n) O_i = sum (i, 1, n) A_i, ie. output total equals input total.
+
+;; GLOSSARY:
+;;  MOD_HASH: this code's sha256 tree hash
+;;  GENESIS_COIN_CHECKER: the function that determines if a coin can mint new cats
+;;  INNER_PUZZLE: an independent puzzle protecting the coins. Solutions to this puzzle are expected to
+;;              generate `AGG_SIG` conditions and possibly `CREATE_COIN` conditions.
+;; ---- items above are curried into the puzzle hash ----
+;;  inner_puzzle_solution: the solution to the inner puzzle
+;;  prev_coin_id: the id for the previous coin
+;;  this_coin_bundle: ((parent_id puzzle_hash amount) . lineage_proof)
+;;  next_coin_bundle: (parent_id inner_puzzle_hash amount)
+;;  prev_subtotal: the subtotal between prev-coin and this-coin
+;;
+;; coin_info: `(parent_id puzzle_hash amount)`. This defines the coin id used with ASSERT_MY_COIN_ID
+;; coin-bundle: the cons box `(coin_info . lineage_proof)`
+;;
+;; and automatically hashed in to the announcement generated with CREATE_COIN_ANNOUNCEMENT.
+;;
+
+(mod (MOD_HASH                 ;; curried into puzzle
+      LIMITATIONS_PROGRAM_HASH ;; curried into puzzle
+      INNER_PUZZLE             ;; curried into puzzle
+      inner_puzzle_solution    ;; if invalid, INNER_PUZZLE will fail
+      limitations_program      ;; The reveal of the hash curried above.  Can be () for default lineage checking.
+      prev_coin_id             ;; used in this coin's announcement, prev_coin ASSERT_COIN_ANNOUNCEMENT will fail if wrong
+      this_coin_bundle         ;; verified with ASSERT_MY_COIN_ID
+      next_coin_bundle         ;; used to generate ASSERT_COIN_ANNOUNCEMENT
+      prev_subtotal            ;; included in announcement, prev_coin ASSERT_COIN_ANNOUNCEMENT will fail if wrong
+      extra_delta              ;; this is the number of "legal discrepancy" between your real delta and what you're announcing your delta is
+    )
+
+     ;;;;; start library code
+
+     (include condition_codes.clvm)
+     (include curry-and-treehash.clinc)
+     (include cat_truths.clib)
+
+     (defmacro assert items
+         (if (r items)
+             (list if (f items) (c assert (r items)) (q . (x)))
+             (f items)
+         )
+     )
+
+     (defun-inline not (arg)
+      (i arg 0 1)
+     )
+
+     ; takes a lisp tree and returns the hash of it
+     (defun sha256tree1 (TREE)
+           (if (l TREE)
+               (sha256 2 (sha256tree1 (f TREE)) (sha256tree1 (r TREE)))
+               (sha256 1 TREE)))
+
+    ; cat_mod_struct = (MOD_HASH MOD_HASH_hash GENESIS_COIN_CHECKER GENESIS_COIN_CHECKER_hash)
+
+    (defun-inline mod_hash_from_cat_mod_struct (cat_mod_struct) (f cat_mod_struct))
+    (defun-inline mod_hash_hash_from_cat_mod_struct (cat_mod_struct) (f (r cat_mod_struct)))
```

Comment:
I believe we no longer need this item in `cat_mod_struct`. It's currently only being used so it can be further curried into children (who only need it so they can pass it to their children, like long-dormant genes in DNA). It's a holdover from the gross hack `sha256tree_with_esc`. Since we're now using the much saner `puzzle-hash-of-curried-function`, we should look into removing this `mod_hash_hash` junk.

Reply:
Upon further review, I think I'm mistaken. The `puzzle-hash-of-curried-function` function expects the curried arguments to have already been tree-hashed. Sorry.